### ISSUE TITLE
KeyboardLayout widget doesn't work as expected

### ISSUE THREAD

**Issue by k4leg** [qtile/qtile#2027]
# Issue description
In `libqtile.widget.keyboardlayout.KeyboardLayout` viewing the current layout is implemented through parsing `setxkbmap` (`setxkbmap -verbose 10 -query`) which is wrong (it doesn't give information about the current layout).
You can check with this example:
```
$ setxkbmap 'us,ru' -option 'grp:caps_toggle'
$ setxkbmap -verbose 10 -query > english # Current layout — English.
$ setxkbmap -verbose 10 -query > русский # Current layout — Russian.
$ diff -Naurp0 english русский # Without output, which means it doesn't change when you change the keyboard layout.
```

`setxkbmap` doesn't let you know the current layout.

Switching layouts also doesn't work as expected. Take a look at an example:
```
# `libqtile.widget.keyboardlayout.KeyboardLayout()`
$ setxkbmap 'us,ru' -option 'grp:caps_toggle,compose:menu' # My setup.
$ setxkbmap us # `libqtile.widget.keyboardlayout.KeyboardLayout` actions when clicking on it.
$ setxkbmap us # `libqtile.widget.keyboardlayout.KeyboardLayout` actions when clicking on it.
# My keyboard layout is now only 'us' I cannot change it back to 'ru' even through the widget, through Caps Lock.

# `libqtile.widget.keyboardlayout.KeyboardLayout(configured_keyboards=['us', 'ru'])`
$ setxkbmap 'us,ru' -option 'grp:caps_toggle,compose:menu' # My setup.
$ setxkbmap ru # `libqtile.widget.keyboardlayout.KeyboardLayout` actions when clicking on it.
$ setxkbmap us # `libqtile.widget.keyboardlayout.KeyboardLayout` actions when clicking on it.
# Thus, I cannot switch the keyboard layout via Caps Lock. Only through the widget even if I add an `option='grp:caps_toggle,compose:menu'` parameter.
```

Also, the keyboard layout doesn't change automatically in the widget if I change it through Caps Lock.

# Qtile version
Since [9bb78e80 ("Adding keyboard layout widget", 2013-08-17)](https://github.com/qtile/qtile/commit/9bb78e80).

**Comment by tych0** [comment]
Can you propose a patch to fix it? I'm not sure I understand all the nuances from your report.

**Comment by k4leg** [comment]
> Can you propose a patch to fix it? I'm not sure I understand all the nuances from your report.

I can implement this using [xkb-switch](https://github.com/grwlf/xkb-switch) or something similar (the easiest way). I think the best option would be to use [python-xlib](https://github.com/python-xlib/python-xlib) but [it doesn't provide the XKB extension](https://github.com/python-xlib/python-xlib/issues/180). I think it's possible to use something like [xkbgroup](https://github.com/hcpl/xkbgroup) but no events are implemented there. I'm new to Python, so I'm not sure if I can implement the third option but I'll try (as a last resort I implement using [xkb-switch](https://github.com/grwlf/xkb-switch) or something similar).

**Comment by tych0** [comment]
We don't use xlib, we use xcb; xcffib needs some patches to provide that too. An xkb-switch implementation would be fine, given that we're already fork()ing for this widget.

**Comment by tcld** [comment]
I am using the `KeyboardLayout`-widget and it works for me, the output of `setxkbmap -verbose 10 -query` correctly returns the current layout. All the other commands you mentioned work fine, too. Which version of Qtile are you using?

**Comment by k4leg** [comment]
> I am using the `KeyboardLayout`-widget and it works for me, the output of `setxkbmap -verbose 10 -query` correctly returns the current layout. All the other commands you mentioned work fine, too. Which version of Qtile are you using?

Qtile version: 3701.d6f8a6b2-1 (AUR, qtile-git). I don't think it depends on the Qtile version.
You have output in
```
$ setxkbmap 'us,ru' -option 'grp:caps_toggle'
$ setxkbmap -verbose 10 -query > a # Current layout — English.
$ setxkbmap -verbose 10 -query > ф # Current layout — Russian.
$ diff -Naurp0 a ф
```
?

**Comment by tcld** [comment]
I have not run the first line, since I have set up my available keymaps in Qtile. But everything else I did run, including things like `setxkbmap gb` to see if everything stays synchronized. And it did - the current layout returned by `setxkbmap -verbose 10 -query` as well as the widget always returned the proper setting.

**Comment by k4leg** [comment]
> I have not run the first line, since I have set up my available keymaps in Qtile. But everything else I did run, including things like `setxkbmap gb` to see if everything stays synchronized. And it did - the current layout returned by `setxkbmap -verbose 10 -query` as well as the widget always returned the proper setting.

Do you have two keyboard layouts? If so, show the output of `setxkbmap -verbose 10 -query` for your two keyboard layouts, please.

**Comment by tych0** [comment]
Can you show the actual output form `setxkbmap -query -verbose 10` instead of a diff?

**Comment by k4leg** [comment]
> Can you show the actual output form `setxkbmap -query -verbose 10` instead of a diff?

```
$ setxkbmap -query -verbose 10
Setting verbose level to 10
locale is C
Trying to load rules file ./rules/evdev...
Trying to load rules file /usr/share/X11/xkb/rules/evdev...
Success.
Applied rules from evdev:
rules:      evdev
model:      pc105
layout:     us,ru
options:    grp:caps_toggle,compose:menu
Trying to build keymap using the following components:
keycodes:   evdev+aliases(qwerty)
types:      complete
compat:     complete
symbols:    pc+us+ru:2+inet(evdev)+capslock(grouplock)+compose(menu)
geometry:   pc(pc105)
rules:      evdev
model:      pc105
layout:     us,ru
options:    grp:caps_toggle,compose:menu
```

**Comment by tych0** [comment]
On Mon, Dec 21, 2020 at 10:22:00AM -0800, k4leg wrote:
> > Can you show the actual output form `setxkbmap -query -verbose 10` instead of a diff?
> 
> ```
> $ setxkbmap -query -verbose 10
> Setting verbose level to 10
> locale is C
> Trying to load rules file ./rules/evdev...
> Trying to load rules file /usr/share/X11/xkb/rules/evdev...
> Success.
> Applied rules from evdev:
> rules:      evdev
> model:      pc105
> layout:     us,ru
> options:    grp:caps_toggle,compose:menu
> Trying to build keymap using the following components:
> keycodes:   evdev+aliases(qwerty)
> types:      complete
> compat:     complete
> symbols:    pc+us+ru:2+inet(evdev)+capslock(grouplock)+compose(menu)
> geometry:   pc(pc105)
> rules:      evdev
> model:      pc105
> layout:     us,ru
> options:    grp:caps_toggle,compose:menu
> ```

Looks to me like `layout: us,ru` is the problem; I think the widget
assumes there's only one thing rendered there.

**Comment by k4leg** [comment]
> Looks to me like `layout: us,ru` is the problem; I think the widget assumes there's only one thing rendered there.

Yes you are right. Perhaps I did not  explain something well (I do not really know English). By the current keyboard layout I mean: "us", "ru", "sk" and others (separately);  not "us,ru,sk". It would also be nice to implement this using events rather than refresh interval. `xkb-switch -W` exactly what is needed (it watches events and prints the current keyboard layout to stdout). But I don't quite understand how to combine this with switching the current keyboard layout (using `xkb-switch -n`).

**Comment by tych0** [comment]
I think in order to use anything besides some command line utility we'll need xkb support in xcffib; I don't remember exactly the details about why it fails to build right now, but I suspect it's nontrival.

I'd definitely takes a patch that ports the widget to use xkb-switch, since it seems like it doesn't really do anything useful right now.

**Comment by ramnes** [comment]
@k4leg Why not just use two distinct layouts and add a Qtile key to switch between the two?

**Comment by k4leg** [comment]
> @k4leg Why not just use two distinct layouts and add a Qtile key to switch between the two?

It will be too crutch. For me for example, switching the current keyboard layout occurs via Caps Lock which will add difficulties. And it will be implemented without watching events. The lag of the current keyboard layout from the displayed one will not look very nice, even if the refresh rate is low.

**Comment by leschalu** [comment]
I think issue is that most shortcuts are not working when only non-latin(greek, cyrillic etc) set with **setxkbmap** so **-layout** option is necessary.

As an option you can update text widget with external script using xkblayout(https://github.com/jsynacek/xkblayout) or similar.

#xkblayout.py body
``` python
from libqtile.command.client import InteractiveCommandClient
from subprocess import run

xkb = run(".config/qtile/xkblayout", shell=True, capture_output=True)
xkb = xkb.stdout.decode()
xkb = xkb.replace('\n', '')

c = InteractiveCommandClient()
c.widget['skbwgt'].update(xkb)
```
#config.py body
``` python
...
from subprocess import run
@hook.subscribe.startup_complete
def startup_complete():
    run('setxkbmap -layout us,gr -option grp:rctrl_rshift_toggle', shell=True)

keys.extend([
        Key(["shift"], 'Control_R', lazy.spawn('python .config/qtile/xkblayout.py')),
        Key(["control"], 'Shift_R', lazy.spawn('python .config/qtile/xkblayout.py'))
])
...
#note name='skbwgt'
                widget.TextBox('us', foreground=bgcol2, background=inactivecol, name='skbwgt', font="liberation mono bold", fontsize=13)
...

**Comment by ep12** [comment]
This was supposed to be re-opened, I think.

**Comment by ep12** [comment]
@tych0 could you remove the stale label? As far as I know this issue is to be kept open until it can be fixed after xcffib supports xcb...

**Comment by AmeerTaweel** [comment]
Any updates on this? I does not seem to be fixed...
